### PR TITLE
refactor: OS discovery files (i - m)

### PR DESCRIPTION
**Pull request by crcro**
#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you signed the [Contributors agreement](http://docs.librenms.org/General/Contributing/)
- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

### CODE REVIEW

File: includes/discovery/os/ibm-amm.inc.php [hunk 1]
```diff
@@ -10,8 +10,6 @@
  * the source code distribution for details.
  */
 
-if (!$os) {
-    if (stristr($sysDescr, 'BladeCenter Advanced Management Module')) {
-        $os = 'ibm-amm';
-    }
+if (str_contains($sysDescr, 'BladeCenter Advanced Management Module')) {
```
Comment (laf) [inline includes/discovery/os/ibm-amm.inc.php]: Needs to be passed true as was stristr before.

File: includes/discovery/os/ibm-imm.inc.php [hunk 2]
```diff
@@ -10,8 +10,6 @@
  * the source code distribution for details.
  */
 
-if (!$os) {
-    if (strstr($sysObjectId, '1.3.6.1.4.1.2.3.51.3')) {
-        $os = 'ibm-imm';
-    }
+if (starts_with($sysObjectId, '.1.3.6.1.4.1.2.3.51.3')) {
```
Comment (laf) [inline includes/discovery/os/ibm-imm.inc.php]: Add this as an elseif rather than separate.

File: includes/discovery/os/ibmnos.inc.php [hunk 3]
```diff
@@ -10,8 +10,6 @@
  * the source code distribution for details.
  */
 
-if (!$os) {
-    if (stristr($sysDescr, 'IBM Networking Operating System') || stristr($sysDescr, 'IBM Flex System Fabric') || stristr($sysDescr, 'IBM Networking OS')) {
-        $os = 'ibmnos';
-    }
+if (str_contains($sysDescr, array('IBM Networking Operating System', 'IBM Flex System Fabric', 'IBM Networking OS'))) {
```
Comment (laf) [inline includes/discovery/os/ibmnos.inc.php]: Again needs passing true as was stristr

File: includes/discovery/os/ibmtl.inc.php [hunk 4]
```diff
@@ -10,9 +10,8 @@
...
 */
 
-if (!$os) {
-    $ibmtl_snmpget = snmp_get($device, 'SML-MIB::product-Name.0', '-Oqv', '');
-    if (stristr($ibmtl_snmpget, 'IBM System Storage TS3500 Tape Library')) {
-        $os = 'ibmtl';
-    }
+$ibmtl_snmpget = snmp_get($device, 'SML-MIB::product-Name.0', '-Oqv', '');
+
+if (str_contains($ibmtl_snmpget, 'IBM System Storage TS3500 Tape Library')) {
```
Comment (laf) [inline includes/discovery/os/ibmtl.inc.php]: Needs passing true as was stristr

File: includes/discovery/os/ios.inc.php [hunk 5]
```diff
@@ -1,32 +1,20 @@
...
-        $os = 'iosxe';
-    }
+if (str_contains($sysDescr, array('Cisco Internetwork Operating System Software', 'IOS (tm)', 'Cisco IOS Software', 'Global Site Selector'))) {
+    $os = 'ios';
+}
 
-    if (strstr($sysDescr, 'IOS XR')) {
-        $os = 'iosxr';
-    }
+if (str_contains($sysDescr, 'IOS-XE')) {
```
Comment (laf) [inline includes/discovery/os/ios.inc.php]: These next two need splitting out to their own OS files.

File: includes/discovery/os/jetdirect.inc.php [hunk 6]
```diff
@@ -1,13 +1,9 @@
...
-        $os = 'jetdirect';
-    }
-    if (strstr($sysObjectId, '.1.3.6.1.4.1.11.1')) {
-        $os = 'jetdirect';
-    }
+if (str_contains($sysDescr, array('JETDIRECT', 'HP ETHERNET MULTI-ENVIRONMENT'))) {
+    $os = 'jetdirect';
+}
+
+if (starts_with($sysObjectId, '.1.3.6.1.4.1.11.1')) {
```
Comment (laf) [inline includes/discovery/os/jetdirect.inc.php]: should be an elseif

File: includes/discovery/os/junos.inc.php [hunk 7]
```diff
@@ -1,9 +1,9 @@
...
-    if (strstr($sysObjectId, '.1.3.6.1.4.1.2636')) {
-        $os = 'junos';
-    } elseif (stristr($sysDescr, 'kernel JUNOS')) {
-        $os = 'junos';
-    }
+if (starts_with($sysObjectId, '.1.3.6.1.4.1.2636')) {
+    $os = 'junos';
+}
+
+if (str_contains($sysDescr, 'kernel JUNOS')) {
```
Comment (laf) [inline includes/discovery/os/junos.inc.php]: Needs to be passed true + should also be elseif

File: includes/discovery/os/ibm-imm.inc.php (hunk 2 above)
Comment (crcro) [inline includes/discovery/os/ibm-imm.inc.php]: elseif to ibm-amm and remove the ibm-imm ?

File: includes/discovery/os/lenovoemc.inc.php [hunk 8]
```diff
@@ -1,15 +1,15 @@
...
 
-if (!$os) {
-    if (strstr($sysDescr, 'EMC SOHO-NAS Storage.')) {
-        $os = 'lenovoemc';
+if (str_contains($sysDescr, 'EMC SOHO-NAS Storage.')) {
+    $os = 'lenovoemc';
+}
 
-        $lenovoemc_mibs = array(
+if (!empty($os)) {
```
Comment (laf) [inline includes/discovery/os/lenovoemc.inc.php]: May as well just have the register mib in the if above.

File: includes/discovery/os/mellanox.inc.php [hunk 9]
```diff
@@ -1,7 +1,5 @@
 <?php
 
-if (!$os) {
-    if (strstr($sysObjectId, ".33049.1.1.1.")) {
-        $os = "mellanox";
-    }
+if (str_contains($sysObjectId, '.33049.1.1.1.')) {
```
Comment (laf) [inline includes/discovery/os/mellanox.inc.php]: Should be .1.3.6.1.4.1.33049.1.1.1.

File: includes/discovery/os/ibm-imm.inc.php (hunk 2 above)
Comment (laf) [inline includes/discovery/os/ibm-imm.inc.php]: Sorry on github it looked like they were the same file :( Ignore that.

**Comment by laf** [comment]
Just needs the Travis issues fixing and all good.

**Comment by crcro** [comment]
i think that there is something wrong with cisco ios xe as it is not containing any IOS-XE in sysDescr:

SNMPv2-MIB::sysDescr.0 = STRING: Cisco IOS Software, ASR1000 Software (X86_64_LINUX_IOSD-UNIVERSALK9-M), Version 15.5(3)S, RELEASE SOFTWARE (fc6)

File: includes/discovery/os/ios.inc.php [hunk 10]
```diff
@@ -2,9 +2,7 @@
 
 if (str_contains($sysDescr, array('Cisco Internetwork Operating System Software', 'IOS (tm)', 'Cisco IOS Software', 'Global Site Selector'))) {
     $os = 'ios';
-}
 
-if (!empty($os)) {
     $extra_mibs = array(
```
Comment (laf) [inline includes/discovery/os/ios.inc.php]: Do these need copying to the other IOS files as well now?

**Comment by laf** [comment]
We have `Cisco IOS Software, IOS-XE Software (PPC_LINUX_IOSD-ADVENTERPRISEK9-M), Version 15.1(3)S, RELEASE SOFTWARE (fc1) Technical Support: http://www.cisco.com/techsupport Copyright (c) 1986-2011 by Cisco Systems, Inc. Compiled Thu 21-Jul-11 21:59 by mcpre` this on ours so IOS-XE seems to be ok

File: includes/discovery/os/iosxe.inc.php [hunk 11]
```diff
@@ -1,5 +1,5 @@
 <?php
 
-if (str_contains($sysDescr, 'IOS-XE')) {
+if (str_contains($sysDescr, array('IOS-XE', 'X86_64_LINUX_IOSD'))) {
```
Comment (laf) [inline includes/discovery/os/iosxe.inc.php]: We've not had reports that IOS-XE hasn't been working. Not sure we should be including this in until we do - are you seeing that sysDescr for IOS-XE based devices?

Comment (crcro) [inline includes/discovery/os/iosxe.inc.php]: yes, that output is from a live ASR1001 equipment of mine

**Comment by laf** [comment]
Odd. Did it not show up as iosxe before?

**Comment by crcro** [comment]
i don't have it added into librenms, just tested it now as a result of checks failure

**Comment by laf** [comment]
Just pushed the fix to exclude IOS-XE from IOS checks.

**Pull request merged**